### DISCUSSION TITLE
migrate: duplicate column name: id in services (panic)

### DISCUSSION THREAD

**Discussion by causa-prima** [photoprism/photoprism#4498] [Q&A] [answered]
I run photoprism on my TrueNAS Scale installation, where I previously used the TrueCharts version that used MariaDB, but now migrated to the official TrueNAS version that uses SQLite. As I already deleted my old containers before I noticed that the backends differ, I didn't use the migration process described in https://docs.photoprism.app/getting-started/advanced/migrations/mariadb-to-sqlite/, but ran the `mysqlite` script on a Linux machine on the MariaDB-backup-sql-file with the following command: `./mysql2sqlite photoprism_backup_20240902.sql > photoprism_backup_20240902_sqlite.sql`. Next, I used photoprism's built-in restore functionality with the created backup file, which ran fine except for some issue with some seemingly duplicated column:
```
\u@:\w$ photoprism restore -i photoprism_backup_20240902_sqlite.sql
INFO[2024-09-02T09:49:54Z] Become a member today, support our mission and enjoy our member benefits! 💎 
INFO[2024-09-02T09:49:54Z] Visit https://www.photoprism.app/membership to learn more. 
INFO[2024-09-02T09:49:54Z] restore: dropping existing sqlite database tables 
INFO[2024-09-02T09:49:54Z] restore: restoring database backup from photoprism_backup_20240902_sqlite.sql 
memory
INFO[2024-09-02T09:50:01Z] restore: index database successfully restored 
INFO[2024-09-02T09:50:01Z] restore: migrating index database schema     
ERRO[2024-09-02T09:50:01Z] migrate: duplicate column name: id (create migrations table) 
ERRO[2024-09-02T09:50:02Z] migrate: failed migrating services           
ERRO[2024-09-02T09:50:02Z] migrate: duplicate column name: id in services (panic) 
WARN[2024-09-02T09:50:02Z] config: password required to initialize 'admin' account 
INFO[2024-09-02T09:50:02Z] completed in 8.058662642s
```
As the migration guide states that you might have to run some migrations manually, I wanted to check the migration status, but again, got the same error when listing the migrations:
```
\u@:\w$ photoprism migrations ls
INFO[2024-09-02T09:52:28Z] Become a member today, support our mission and enjoy our member benefits! 💎 
INFO[2024-09-02T09:52:28Z] Visit https://www.photoprism.app/membership to learn more. 
ERRO[2024-09-02T09:52:28Z] migrate: duplicate column name: id (create migrations table)
```
Just running the failed migrations doesn't work as well:
```
\u@:\w$ photoprism migrations run -f
INFO[2024-09-02T10:09:25Z] Become a member today, support our mission and enjoy our member benefits! 💎 
INFO[2024-09-02T10:09:25Z] Visit https://www.photoprism.app/membership to learn more. 
INFO[2024-09-02T10:09:25Z] migrate: running previously failed migrations 
INFO[2024-09-02T10:09:25Z] migrating database schema...                 
ERRO[2024-09-02T10:09:25Z] migrate: duplicate column name: id (create migrations table) 
ERRO[2024-09-02T10:09:26Z] migrate: failed migrating migrations         
ERRO[2024-09-02T10:09:26Z] migrate: duplicate column name: id in migrations (panic) 
WARN[2024-09-02T10:09:27Z] config: password required to initialize 'admin' account 
INFO[2024-09-02T10:09:27Z] completed in 1.081151517s
```

My initial though from these messages was that the `migrations` or `services` table somehow got two `id` columns, but the `CREATE TABLE` statement in the backup file just have one:
```sql
CREATE TABLE `migrations` (
  `id` varchar(16) NOT NULL
,  `dialect` varchar(16) DEFAULT NULL
,  `error` varchar(255) DEFAULT NULL
,  `source` varchar(16) DEFAULT NULL
,  `started_at` datetime DEFAULT NULL
,  `finished_at` datetime DEFAULT NULL
,  `stage` varchar(16) DEFAULT NULL
,  PRIMARY KEY (`id`)
);

CREATE TABLE `services` (
  `id` integer  NOT NULL PRIMARY KEY AUTOINCREMENT
,  `acc_name` varchar(160) DEFAULT NULL
,  `acc_owner` varchar(160) DEFAULT NULL
,  `acc_url` varchar(255) DEFAULT NULL
,  `acc_type` varbinary(255) DEFAULT NULL
,  `acc_key` varbinary(255) DEFAULT NULL
,  `acc_user` varbinary(255) DEFAULT NULL
,  `acc_pass` varbinary(255) DEFAULT NULL
,  `acc_error` varbinary(512) DEFAULT NULL
,  `acc_errors` integer DEFAULT NULL
,  `acc_share` integer DEFAULT NULL
,  `acc_sync` integer DEFAULT NULL
,  `retry_limit` integer DEFAULT NULL
,  `share_path` varbinary(500) DEFAULT NULL
,  `share_size` varbinary(16) DEFAULT NULL
,  `share_expires` integer DEFAULT NULL
,  `sync_path` varbinary(500) DEFAULT NULL
,  `sync_status` varbinary(16) DEFAULT NULL
,  `sync_interval` integer DEFAULT NULL
,  `sync_date` datetime DEFAULT NULL
,  `sync_upload` integer DEFAULT NULL
,  `sync_download` integer DEFAULT NULL
,  `sync_filenames` integer DEFAULT NULL
,  `sync_raw` integer DEFAULT NULL
,  `created_at` datetime DEFAULT NULL
,  `updated_at` datetime DEFAULT NULL
,  `deleted_at` datetime DEFAULT NULL
,  `acc_timeout` varbinary(16) DEFAULT NULL
);
```
I also looked at all the other `CREATE` statements, but couldn't find any with two `id` columns.

However, despite these error messages, photoprism looks just as before the migration, as I can see all the pictures and albums I was expecting. Still, these error messages worry me, because they might lead to failing migrations in the future. Any hints on how I could debug this further or what I should do?

**Comment by causa-prima** [comment]
I just noticed I get similar messages in the log when starting up the container:

```
2024-09-02 13:45:03.502722+02:00time="2024-09-02T11:45:03Z" level=error msg="migrate: duplicate column name: id (create migrations table)"
2024-09-02 13:45:04.505636+02:00time="2024-09-02T11:45:04Z" level=error msg="migrate: failed migrating passwords"
2024-09-02 13:45:04.506402+02:00time="2024-09-02T11:45:04Z" level=error msg="migrate: duplicate column name: uid in passwords (panic)"
```

**Comment by causa-prima** [comment]
I noticed that I used some bash script with the same name to convert the SQL file created by `photoprism backup` that is different from the python script stated in the migration guide. Thus, I created a new mariadb database with the sql script created by photoprism and called the python script on it to create the SQLite index.db file. After startup of photoprism, I got these warnings:

```
2024-09-03 12:18:39.728058+02:00time="2024-09-03T10:18:39Z" level=error msg="migrate: failed migrating albums"
2024-09-03 12:18:39.728166+02:00time="2024-09-03T10:18:39Z" level=error msg="migrate: index idx_albums_album_slug already exists in albums (panic)"
```
The already-existing index doesn't worry me, but some failing migration on the `allbums` table might be trouble later. Too bad the message doesn't state what failed - is there a way to debug that?

After this process, I now get this messages periodically (probably every time that photoprism does its maintainance work):
```
2024-09-03 12:19:48.303019+02:00time="2024-09-03T10:19:48Z" level=warning msg="file fsits22enw5r55gr: sql: Scan error on column index 11, name \"embeddings_json\": unsupported Scan, storing driver.Value type string into type *json.RawMessage while loading markers"
```
The message seems to prevent new markers from being created. My guess is that some kind of data transformation needs to be performed on the `embeddings_json` column. Any hints about that?

**Comment by graciousgrey** [comment]
Unfortunately, I can't help much with the migration. Maybe another community member has more experience with this and can help.

If you haven't already done so, it might be worth checking if it's possible to use mariadb instead of sqlite on TrueNAS Scale before you put more work into the migration. Sqlite is only recommended for small libraries. So even if the migration works, there could be performance issues.

https://docs.photoprism.app/getting-started/faq/#should-i-use-sqlite-mariadb-or-mysql
https://docs.photoprism.app/getting-started/troubleshooting/sqlite/

**Comment by causa-prima** [comment, marked as answer]
After I spend even more work on this by trying to copy the old data to the new schema and getting even other problems with that, I finally decided to create a mariadb container as suggested, restored my mariadb-backup to that one and am not getting any strange error messages as before anymore.
